### DISCUSSION TITLE
Proper way of using Scene & Game<Scene> and using ttf font

### DISCUSSION THREAD

**Discussion by Its-Kenta** [littlektframework/littlekt#149] [Q&A] [answered]
Hi,

I'm rather confused by how to properly use Game<Scene> & Scene and I can't seem to find much of this in the samples, or perhaps there is, but all the code being bunched together makes it difficult to read it imho. I like to separate my code to maintain clean workspace and I would like to separate my scenes from MainMenuScene and PlayScene.

- The main entry point should be Game<Scene>  correct?
- Normal Scene can't load assets (if I understand correctly), that means I have to load *all* of my assets in the main entry point and not the scenes respectively?
- How would I go about changing scenes from one scene to another? (e.g. From PlayScene to PauseScene and vice versa?)

I understand that there is still no documentation on how to load TTF, but is there any short info on how to exactly draw text using TTF font? Or if not, how would I go about converting it to .FNT bitmap?

Thanks!

**Comment by LeHaine** [comment, marked as answer]
Hey!

Th `Game` class is an optional class that contains a bunch of tidbits for handling and changing scenes. The `Game` class is completely optional and does not have to be used. 

> The main entry point should be Game correct?

As mentioned above it is completely optional. The only thing that needs to be passed in when creating the app is a `ContextListener`. The `Game` class is a `ContextListener`.

> Normal Scene can't load assets (if I understand correctly), that means I have to load all of my assets in the main entry point and not the scenes respectively?

Not necessarily.  A `Scene` has a reference to the `Context` when it's created. With `Context` we can access the `resourcesVfs: VfsFile` which we can use to load assets. We can load assets directly in a `Scene` when it's created by making us of the `prepare` delegate or we can use an [AssetProvider](https://littlekt.com/docs/io/assets) which does a lot of the work for us.

> How would I go about changing scenes from one scene to another? (e.g. From PlayScene to PauseScene and vice versa?)

We need to add those scenes to the `Game` instance and then set the scene:

```kotlin
class SampleGame(context: Context) : Game<Scene>(context) {

    override suspend fun Context.start() {
        setSceneCallbacks(this) // need to call this to subscribe to the scene callbacks in Game class

        // register the scenes - only need to do it once per scene. We can also register it later if we want.
        addScene(PlayScene())
        addScene(MenuScene())

        setScene<MenuScene>() // set initial scene - pass in by class type

        onRender {
            if (input.isKeyJustPressed(Key.BACKSPACE)) {
                if (currentScene !is MenuScene) {
                    setScene<MenuScene>()
                }
            }
        }
    }
}
```

> I understand that there is still no documentation on how to load TTF, but is there any short info on how to exactly draw text using TTF font? 

While we can load a `TtfFont` just fine there are still issues with rendering it with `GpuFont`. See issue #23.

Loading a Ttf: 
```kotlin
val font: TtfFont = resourcesVfs["arial.ttf"].readTtfFont()
```

Using a `GpuFont`:
```kotlin
val batch = SpriteBatch(context)
val gpuFont = GpuFont(context, font)
gpuFont.addText("My Text!", x = 50, y = 100, pxSize = 16)

onRender {
    batch.use {
        gpuFont.useShaderWith(it) // need to set to GpuFont shader
        gpuFont.draw(it)
    }
    batch.shader = batch.defaultShader // set back to default SpriteBatch shader
}
```

Using a `BitmapFont` works great as well and has a lot more [documentation](https://littlekt.com/docs/2d/fonts/bitmap-fonts) for it. I personally use [BMFont](https://www.angelcode.com/products/bmfont/) to convert a TTF to a Bitmap. But really any bitmap font generator can be used.

**Comment by Its-Kenta** [comment]
Thank you for the further info! That makes a lot more sense!
